test/integration/test_agent_sink.go

```diff
@@ -0,0 +1,202 @@
+package e2e
+
+import (
+	"encoding/json"
+	"io/ioutil"
+	"net/http"
+	"net/http/httptest"
+	"reflect"
+	"strings"
+	"sync"
+	"time"
+
+	sdkEvent "github.com/newrelic/infra-integrations-sdk/data/event"
+	"github.com/newrelic/nri-kube-events/pkg/events"
+	"github.com/newrelic/nri-kube-events/pkg/sinks"
+	log "github.com/sirupsen/logrus"
+)
+
+// Must be in sync with unexported name in pkg/sinks/new_relic_infra.go:32.
+const newRelicInfraSinkID = "newRelicInfra"
+
+// TestAgentSink is an instrumented infra-agent sink for testing e2e reception and processing.
+type TestAgentSink struct {
+	agentSink         events.Sink
+	httpServer        *httptest.Server
+	eventReceivedChan chan struct{}
+	receivedEvents    []sdkEvent.Event
+	mtx               *sync.RWMutex
+}
+
+// NewTestAgentSink returns an instrumented infra-agent sink for testing.
+func NewTestAgentSink() *TestAgentSink {
+	mockedAgentSink := &TestAgentSink{
+		mtx:               &sync.RWMutex{},
+		eventReceivedChan: make(chan struct{}, 128),
+	}
+	mockedAgentSink.httpServer = httptest.NewServer(mockedAgentSink)
+
+	agentSinkConfig := sinks.SinkConfig{
+		Name: newRelicInfraSinkID,
+		Config: map[string]string{
+			"clusterName":   "integrationTest",
+			"agentEndpoint": "http://" + mockedAgentSink.httpServer.Listener.Addr().String(),
+		},
+	}
+
+	createdSinks, err := sinks.CreateSinks([]sinks.SinkConfig{agentSinkConfig})
+	if err != nil {
+		log.Fatalf("error creating infra sink: %v", err)
```

Comment:
Why not return error instead of halting entire process?

Reply:
Or maybe have `testing.T` as field of `TestAgentSink`, so also `ServeHTTP` can call `t.Fatalf` ?